### PR TITLE
feat(output): add Sarif / Gitlab support

### LINKED ISSUE #982: Add support for SARIF format

### PR DESCRIPTION
**Pull request by cfabianski**
## Description

Fixes https://github.com/Bearer/bearer/issues/982

Add extra format for security output (`sarif` for GitHub and `gitlab-sast` for GitLab)

**GitLab**

The configuration would look like this

```yaml
image:
  name: bearer/bearer
  entrypoint: [ "" ]

bearer:   
  script:     
    - bearer scan . --format gitlab-sast --output gl-sast-report.json    
  
  artifacts:    
    reports:       
      sast: gl-sast-report.json
```

**GitHub**

```yaml
name: Bearer

on:
  push:
    branches:
      - main

permissions:
  contents: read
  security-events: write

jobs:
  rule_check:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: Run Report
        id: report
        uses: bearer/bearer-action@v2
        with:
          # Check that those 2 options are available
          format: sarif
          output: results.sarif
      - name: Upload SARIF file
        uses: github/codeql-action/upload-sarif@v2
        with:
          sarif_file: results.sarif
```

Also done as part of this PR

- Separate Filename and FullFilename (FullFilename is mainly used for the terminal output to get a direct access to the file)
  - Compute Fingerprint based on Filename 
  - Kept the previous version of the Fingerprint which was using FullFilename as OldFingerprint (`old_fingerprint` in the JSON output)
- Introduce `source` and `sink` nodes so that we have a more detailed position report

## Checklist

- [ ] I've added test coverage that shows my fix or feature works as expected.
- [ ] I've updated or added documentation if required.
- [ ] I've included usage information in the description if CLI behavior was updated or added.
- [ ] PR title follows [Conventional Commits](https://www.conventionalcommits.org/) format

### CODE REVIEW

**Comment by markmichon** [comment]
Please include updates to the CLI commands and flags for docs:
```
go run ./scripts/gen-doc-yaml.go
```

File: new/detector/composition/composition.go [hunk 1]
```diff
@@ -23,18 +23,24 @@ func ReportDetections(report reportdetections.ReportDetection, file *file.FileIn
 		data := detection.Data.(custom.Data)
 
 		if len(data.Datatypes) == 0 {
+			// log.Error().Msgf("%s %s %d %d", file.Path, detection.MatchNode.Content(), detection.MatchNode.StartLineNumber(), detection.MatchNode.StartColumnNumber())
```
Comment (elsapet) [inline new/detector/composition/composition.go]:
```suggestion
```

**Review by elsapet — APPROVED** [review]

File: new/detector/composition/java/.snapshots/flow/TestFlow--different-line.yml [hunk 2]
```diff
@@ -6,11 +6,28 @@ high:
         description: ""
         documentation_url: ""
       line_number: 2
+      full_filename: different-line.java
       filename: different-line.java
       category_groups:
         - PII
         - Personal Data
+      source:
```
Comment (didroe) [inline new/detector/composition/java/.snapshots/flow/TestFlow--different-line.yml:14]: Meaning of source+sink should be swapped

File: e2e/flags/.snapshots/TestReportFlagsShouldFail-invalid-format-flag [hunk 3]
```diff
@@ -1,6 +1,6 @@
 
 --
-Error: flag error: report flags error: invalid format argument; supported values: json, yaml
+Error: flag error: report flags error: invalid format argument; supported values: json, yaml, sarif
```
Comment (didroe) [inline e2e/flags/.snapshots/TestReportFlagsShouldFail-invalid-format-flag]: Missing `gitlab-sast` here

**Comment by cfabianski** [comment]
@didroe can I get another review please?

**Review by didroe — APPROVED** [review]

**Pull request merged**